### DISCUSSION TITLE
Custom mutation hook

### DISCUSSION THREAD

**Discussion by chertik77** [TanStack/query#7547] [Q&A]
Hi, i want to figure out how i can infer return type of mutation function.

I have a signin function , it returns type AuthResponse.

```typescript
 async signin(data: SigninSchema) {
    const response = await axiosInstance.post<AuthResponse>(
      ApiEndpoints.Signin,
      data
    )

    return response.data
  },
```

i have a custom useAppMuattion hook 

```typescript
export const useAppMutation = <T = void, R = void>({
  mutationKey,
  mutationFn,
  toastErrorMessage,
  ...options
}: UseMutationOptions<R, AxiosError, T> & UseAppMutationProps) =>
  useMutation({
    mutationKey,
    mutationFn,
    onError() {
      toast.error(toastErrorMessage)
    },
    ...options
  })
```
and here is usage of this hook: 

```typescript
return useAppMutation<SigninSchema, AuthResponse>({
    mutationKey: ['signin'],
    mutationFn: data => authService.signin(data),
    onSuccess(data) {
      reset()
      dispatch(authenticate(data))
    },
    onError: e =>
      toast.error(
        e.response?.status === 401
          ? 'Sign-in failed. The email or password you entered is incorrect. Please try again.'
          : 'An error occurred during sign-in. Our technical team has been notified. Please try again shortly.'
      )
  })

```

but my axios function already returning type AuthResponse, and here in signin hook, i need to pass in again, i tried ReturnType<typeof mutationFn> it doesnt work, maybe someone have any ideas to work around it ?

**Comment by serdar-webdev** [comment]
try this: 
const res = axios.post<{}, AxiosResponse<ResBody>, ReqBody>

res.data <--- will be ResBody

**Reply by chertik77** [reply]
> try this: const res = axios.post<{}, AxiosResponse, ReqBody>
> 
> res.data <--- will be ResBody

U didnt get me, I dont want to pass the type to the custom hook, cause axios func already returning it

**Reply by serdar-webdev** [reply]
You're right, i didin't get you, can you tell more about your problem, maybe i can help you
